lib/Optimizer/Transforms/DecompositionPatterns.cpp

```diff
@@ -0,0 +1,922 @@
+struct TToPhasedRx : public OpRewritePattern<quake::TOp> {
+  using OpRewritePattern<quake::TOp>::OpRewritePattern;
+
+  void initialize() { setDebugName("TToPhasedRx"); }
+
+  LogicalResult matchAndRewrite(quake::TOp op,
+                                PatternRewriter &rewriter) const override {
+    if (!op.getControls().empty())
+      return failure();
+    if (!quake::isAllReferences(op))
+      return failure();
+
+    // Op info
+    Location loc = op->getLoc();
+    Value target = op.getTarget();
+    Value angle = createConstant(loc, -M_PI_4, rewriter.getF64Type(), rewriter);
+    if (op.isAdj())
+      angle = rewriter.create<arith::NegFOp>(loc, angle);
+
+    // Necessary/Helpful constants
+    ValueRange noControls;
+    Value zero = createConstant(loc, 0.0, rewriter.getF64Type(), rewriter);
+    Value pi_2 = createConstant(loc, M_PI_2, rewriter.getF64Type(), rewriter);
+    Value negPi_2 = rewriter.create<arith::NegFOp>(loc, pi_2);
+
+    std::array<Value, 2> parameters = {pi_2, zero};
+    rewriter.create<quake::PhasedRxOp>(loc, parameters, noControls, target);
+    parameters[0] = angle;
+    parameters[1] = pi_2;
+    rewriter.create<quake::PhasedRxOp>(loc, parameters, noControls, target);
+    parameters[0] = negPi_2;
+    parameters[1] = zero;
+    rewriter.create<quake::PhasedRxOp>(loc, parameters, noControls, target);
+
+    rewriter.eraseOp(op);
+    return success();
+  }
+};
+
+//===----------------------------------------------------------------------===//
+// XOp decompositions
+//===----------------------------------------------------------------------===//
+
+struct CXToCZ : public OpRewritePattern<quake::XOp> {
+  using OpRewritePattern<quake::XOp>::OpRewritePattern;
+
+  void initialize() { setDebugName("CXToCZ"); }
+
+  LogicalResult matchAndRewrite(quake::XOp op,
+                                PatternRewriter &rewriter) const override {
+    if (!quake::isAllReferences(op))
+      return failure();
+    if (failed(checkNumControls(op, 1)))
+      return failure();
+
+    // Op info
+    Location loc = op->getLoc();
+    Value target = op.getTarget();
+
+    rewriter.create<quake::HOp>(loc, target);
+    rewriter.create<quake::ZOp>(loc, op.getControls(), target);
+    rewriter.create<quake::HOp>(loc, target);
+    rewriter.eraseOp(op);
+    return success();
+  }
+};
+
+struct CCXToCCZ : public OpRewritePattern<quake::XOp> {
+  using OpRewritePattern<quake::XOp>::OpRewritePattern;
+
+  void initialize() { setDebugName("CCXToCCZ"); }
+
+  LogicalResult matchAndRewrite(quake::XOp op,
+                                PatternRewriter &rewriter) const override {
+    if (!quake::isAllReferences(op))
+      return failure();
+    if (failed(checkNumControls(op, 2)))
+      return failure();
+
+    // Op info
+    Location loc = op->getLoc();
+    Value target = op.getTarget();
+
+    rewriter.create<quake::HOp>(loc, target);
+    rewriter.create<quake::ZOp>(loc, op.getControls(), target);
```

Comment:
I'm guessing that any negated controls are carried through here, and then handled later in something like `CZToCX`?